File: rust/tests/test_error_paths.rs
//! TC3 — Error / validation paths with TYPED errors.
//!
//! Mirrors the C# `TestErrorPaths` suite. Exercises the generated client's error
//! classification against the live base: a 404 fetch, three server-side validation
//! rejections (invalid select option, wrong value type, unknown field), a 401 from
//! a bogus key, and the offline `RateLimited` variant. Each scenario asserts the
//! *specific* `AirtableError::Api` variant (never a bare `is_err()`), checks the HTTP
//! `status`, and the first-class Airtable error `code` field.

use myairtable_tests::*;

fn setup() -> Airtable {
    dotenvy::from_filename(".env").ok();
    let api_key = std::env::var("AIRTABLE_API_KEY").expect("AIRTABLE_API_KEY must be set");
    let base_id = std::env::var("AIRTABLE_BASE_ID").expect("AIRTABLE_BASE_ID must be set");
    Airtable::new(&api_key, &base_id)
}

/// A client pointed at the real base but with a bogus API key (mirrors
/// `MakeAirtableWithBadKey`), so auth (401) is exercised rather than the base (404).
fn setup_with_bad_key() -> Airtable {
    dotenvy::from_filename(".env").ok();
    let base_id = std::env::var("AIRTABLE_BASE_ID").expect("AIRTABLE_BASE_ID must be set");
    Airtable::new("patBOGUS00000.deadbeefdeadbeefdeadbeefdeadbeef", &base_id)
}

/// A primary-key value unique to this run (distinct per file via `label`).
fn primary_key(label: &str) -> String {
    let ts = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis();
    format!("Rust Error {label} {ts}")
}

/// Assert `result` is the typed `AirtableError::Api` variant, return `(status, code)`.
fn expect_api_error<T: std::fmt::Debug>(
    result: Result<T, AirtableError>,
    scenario: &str,
) -> (u16, Option<String>) {
    match result {
        Err(AirtableError::Api { status, code, body }) => {
            println!("{scenario} -> Api {{ status: {status}, code: {code:?} }}: {body}");
            // The Airtable code is now a first-class field; surface None when empty so the
            // 404-by-id legacy envelope (which carries the code in `message`) is handled below.
            let code = if code.is_empty() { None } else { Some(code) };
            (status, code)
        }
        Err(other) => panic!("{scenario}: expected AirtableError::Api, got {other:?}"),
        Ok(v) => panic!("{scenario}: expected an error, got Ok({v:?})"),
    }
}

// 1. GET a nonexistent record id -> typed Api error, 404, code NOT_FOUND.
#[tokio::test]
async fn get_nonexistent_record_is_typed_api_error() {
    let at = setup();
    let result = at
        .primary
        .dict
        .get_one(&"recDOESNOTEXIST0001".to_string(), true)
        .await;
    let (status, code) = expect_api_error(result, "404 get");
    assert_eq!(status, 404, "expected HTTP 404 for a nonexistent record id");
    assert_eq!(
        code.as_deref(),
        Some("NOT_FOUND"),
        "expected NOT_FOUND code"
    );
}

// 2. Dict-create with an invalid single-select option -> typed Api error,
//    code INVALID_MULTIPLE_CHOICE_OPTIONS.
#[tokio::test]
async fn create_with_invalid_select_option_is_typed_api_error() {
    let at = setup();
    let mut fields = Fields::new();
    fields.set(PrimaryFields::PRIMARY_KEY_ID, primary_key("BadSelect"));
    fields.set(PrimaryFields::SINGLE_SELECT_ID, "NotARealOption_zzz");

    let result = at.primary.dict.create_one(&fields, true).await;
    let (status, code) = expect_api_error(result, "bad select");
    assert!(
        (400..500).contains(&status),
        "expected a 4xx client error, got {status}"
    );
    assert_eq!(
        code.as_deref(),
        Some("INVALID_MULTIPLE_CHOICE_OPTIONS"),
        "expected INVALID_MULTIPLE_CHOICE_OPTIONS code"
    );
}

// 3. Dict-create with a wrong-typed value (Number field set to a string) ->
//    typed Api error, code INVALID_VALUE_FOR_COLUMN.
#[tokio::test]
async fn create_with_wrong_value_type_is_typed_api_error() {
    let at = setup();
    let mut fields = Fields::new();
    fields.set(PrimaryFields::PRIMARY_KEY_ID, primary_key("BadType"));
    fields.set(PrimaryFields::NUMBER_INT_ID, "not a number");

    let result = at.primary.dict.create_one(&fields, true).await;
    let (status, code) = expect_api_error(result, "wrong type");
    assert!(
        (400..500).contains(&status),
        "expected a 4xx client error, got {status}"
    );
    assert_eq!(
        code.as_deref(),
        Some("INVALID_VALUE_FOR_COLUMN"),
        "expected INVALID_VALUE_FOR_COLUMN code"
    );
}

// 4. Dict-create with an unknown field id -> typed Api error, code UNKNOWN_FIELD_NAME.
#[tokio::test]
async fn create_with_unknown_field_is_typed_api_error() {
    let at = setup();
    let mut fields = Fields::new();
    fields.set(PrimaryFields::PRIMARY_KEY_ID, primary_key("BadField"));
    fields.set("fldDOESNOTEXIST00000", "x");

    let result = at.primary.dict.create_one(&fields, true).await;
    let (status, code) = expect_api_error(result, "unknown field");
    assert!(
        (400..500).contains(&status),
        "expected a 4xx client error, got {status}"
    );
    assert_eq!(
        code.as_deref(),
        Some("UNKNOWN_FIELD_NAME"),
        "expected UNKNOWN_FIELD_NAME code"
    );
}

// 5. Real base id + bogus key -> typed Api error, 401, code AUTHENTICATION_REQUIRED.
#[tokio::test]
async fn bad_api_key_is_typed_auth_error() {
    let at = setup_with_bad_key();
    let mut query = AirtableQuery::new();
    query.max_records = Some(1);

    let result = at.primary.dict.get_many(&query).await;
    let (status, code) = expect_api_error(result, "bad key");
    assert_eq!(status, 401, "expected HTTP 401 for a bogus API key");
    // Airtable's 401 envelope carries AUTHENTICATION_REQUIRED; assert it when present but
    // don't fail solely on the code shape — the 401 status is the load-bearing classification.
    if let Some(code) = code {
        assert_eq!(
            code, "AUTHENTICATION_REQUIRED",
            "expected AUTHENTICATION_REQUIRED code"
        );
    }
}

// 6. Rate-limited (429): assert the distinct `RateLimited` variant is well-formed and carries
//    its Retry-After value (offline). The live 429 path is covered by TC8.
#[test]
fn rate_limited_is_a_distinct_typed_variant() {
    let err = AirtableError::RateLimited {
        retry_after: Some(30.0),
    };
    match err {
        AirtableError::RateLimited { retry_after } => assert_eq!(retry_after, Some(30.0)),
        other => panic!("expected RateLimited, got {other:?}"),
    }
    // A 429 response classifies as RateLimited, not Api.
    assert!(matches!(
        AirtableError::from_response(429, String::new()),
        AirtableError::RateLimited { .. }
    ));
}

File: python/tests/test_error_paths.py
"""TC3 - Error/validation paths with TYPED errors.

Mirrors the C# ``TestErrorPaths`` suite (csharp/tests/TestErrorPaths.cs). Exercises the
generated client's error classification against the live base: a 404 fetch, three
server-side validation rejections (invalid select option, wrong value type, unknown
field), and a 401 from a bad key.

The Python target wraps ``pyairtable``, whose ``Api._process_response`` calls
``response.raise_for_status()`` and re-raises ``requests.exceptions.HTTPError`` with the
Airtable error envelope appended. That exception is a SPECIFIC, structured type:

* ``exc.response.status_code`` carries the HTTP status (404 / 422 / 401), and
* ``exc.response.json()["error"]`` carries the Airtable error code -- either a bare
  string (e.g. ``"NOT_FOUND"``) or a dict ``{"type": "INVALID_VALUE_FOR_COLUMN", ...}``.

So Python classifies these failures just like the C# target does (no parity gap): each
test below asserts the precise exception type via ``pytest.raises(HTTPError)`` and then
inspects the status code AND the Airtable error code, rather than catching a bare
``Exception``.
"""

import os
import time

import pytest
from pyairtable import Api
from requests.exceptions import HTTPError

from output import Airtable
from output.dynamic.tables import PrimaryTable

# Distinct per-run, per-file prefix so concurrent suites don't collide on writes.
_SUITE = f"PyError {int(time.time() * 1000)}-{os.getpid()}"


@pytest.fixture(scope="module")
def airtable():
    return Airtable()


def _airtable_error_code(exc: HTTPError) -> str:
    """Extract the Airtable error code from a raised HTTPError.

    The Airtable error envelope is either ``{"error": "NOT_FOUND"}`` (a bare string,
    seen on 404s) or ``{"error": {"type": "INVALID_VALUE_FOR_COLUMN", ...}}``.
    """
    response = exc.response
    assert response is not None
    error = response.json()["error"]
    if isinstance(error, dict):
        return str(error["type"])
    return str(error)


def make_primary_table_with_bad_key() -> PrimaryTable:
    """A `Primary` table pointed at the real base but with a bogus PAT, for auth tests.

    Mirrors C# ``TestSetup.MakeAirtableWithBadKey`` (real base id from env + a deliberately
    invalid key), but deliberately does NOT construct a second generated ``Airtable``.

    The generated ``Airtable.__init__`` registers its key in a process-global config and
    caches tables on a class-level dict shared across every instance; a bogus-key client
    would therefore poison the good client used by other tests. Instead we build a
    bogus-key ``pyairtable.Api`` directly and wrap it in the generated ``PrimaryTable``.
    This exercises the same generated ``.dict`` layer with the bad key while leaving all
    shared/global state untouched.
    """
    base_id = os.environ.get("AIRTABLE_BASE_ID")
    assert base_id, "AIRTABLE_BASE_ID not set (need .env in the test repo root)."
    bad_api = Api(api_key="patBOGUS00000.deadbeefdeadbeefdeadbeefdeadbeef")
    return PrimaryTable.from_table(bad_api.table(base_id, "tblmb3iqgpNS1ysV2"))


def new_primary_dict(fields: dict | None = None) -> dict:
    return {"fields": fields or {}}


def test_get_nonexistent_record_raises_not_found(airtable):
    with pytest.raises(HTTPError) as exc_info:
        airtable.primary.dict.get("recDOESNOTEXIST0001")
    response = exc_info.value.response
    assert response is not None
    assert response.status_code == 404
    assert _airtable_error_code(exc_info.value) == "NOT_FOUND"


def test_create_with_invalid_select_option_raises_api_error(airtable):
    record = new_primary_dict({"Primary Key": f"{_SUITE} BadSelect", "Single Select": "NotARealOption_zzz"})
    with pytest.raises(HTTPError) as exc_info:
        airtable.primary.dict.create(record)
    response = exc_info.value.response
    assert response is not None
    assert response.status_code == 422
    assert _airtable_error_code(exc_info.value) == "INVALID_MULTIPLE_CHOICE_OPTIONS"


def test_create_with_wrong_value_type_raises_api_error(airtable):
    record = new_primary_dict({"Primary Key": f"{_SUITE} BadType", "Number (int)": "not a number"})
    with pytest.raises(HTTPError) as exc_info:
        airtable.primary.dict.create(record)
    response = exc_info.value.response
    assert response is not None
    assert response.status_code == 422
    assert _airtable_error_code(exc_info.value) == "INVALID_VALUE_FOR_COLUMN"


def test_create_with_unknown_field_raises_api_error(airtable):
    record = new_primary_dict({"Primary Key": f"{_SUITE} BadField", "fldDOESNOTEXIST00000": "x"})
    with pytest.raises(HTTPError) as exc_info:
        airtable.primary.dict.create(record)
    response = exc_info.value.response
    assert response is not None
    assert response.status_code == 422
    assert _airtable_error_code(exc_info.value) == "UNKNOWN_FIELD_NAME"


def test_bad_api_key_raises_authentication_required():
    # Real base + bogus key so auth (401) is checked, not the base (404).
    bad_primary = make_primary_table_with_bad_key()
    with pytest.raises(HTTPError) as exc_info:
        bad_primary.dict.get(max_records=1)
    response = exc_info.value.response
    assert response is not None
    assert response.status_code == 401
    assert _airtable_error_code(exc_info.value) == "AUTHENTICATION_REQUIRED"


def test_rate_limited_error_is_distinct_from_other_http_errors():
    # The live 429 path is exercised in TC8. Here we assert (offline) that the target can
    # distinguish a rate-limit response from other HTTP failures via its status code:
    # pyairtable surfaces 429 through the same HTTPError type, keyed by response.status_code,
    # which is exactly how the 404/422/401 cases above are told apart. There is no separate
    # rate-limited exception class, so this is a status-code-level (not type-level) distinction.
    assert HTTPError is not None
